autoload/suda.vim

```diff
@@ -199,6 +198,48 @@ function! suda#FileWriteCmd() abort
   endtry
 endfunction
 
+let s:smart_read_checked = {}
+function! suda#BufEnter() abort
+  let bufname = expand('<afile>')
+  let bufnr = bufnr('%')
+  if get(s:smart_read_checked, bufnr, 0)
+    " returning because bufnr was found in cache dict
+    return
+  else
+    " updating cache dict
+    let s:smart_read_checked[bufnr] = 1
+  end
+  if !empty(&buftype) || isdirectory(bufname) || empty(bufname) || match(bufname, '^[a-z]\+://*') isnot# -1
+    " Non file buffer
+    return
+  elseif filereadable(bufname) && filewritable(bufname)
+    " File is readable and writeable
+    return
+  elseif empty(getftype(bufname))
+    " if file doesn't exist, we search for a all directories up it's path to
+    " see if each one of them is writable, if not, we `return`
+    let parent = fnamemodify(bufname, ':p')
+    while parent !=# fnamemodify(parent, ':h')  " stop loop on root
+      let parent = fnamemodify(parent, ':h')
+      if filewritable(parent) is# 2
+        " reached a writeable directory - no need to switch to suda://
+        return
+      elseif !filereadable(parent) && isdirectory(parent)
+        " reached a readable parent directory which is not writeable so it
+        " means we finally can be sure we need to replace our buffer with a
+        " suda:// .
+        break
+      endif
+    endwhile
+  endif
+  execute printf(
+        \ 'keepalt keepjumps edit %s%s',
+        \ g:suda#prefix,
+        \ fnamemodify(bufname, ':p'),
+        \)
+  execute printf('silent! %dbwipeout', bufnr)
+endfunction
```

Comment:
**must**

Please replace the function to the code below and confirm the behavior

```vim
function! suda#BufEnter() abort
  if exists('b:suda_smart_edit_checked')
    return
  endif
  let b:suda_smart_edit_checked = 1
  let bufname = expand('<afile>')
  if !empty(&buftype)
        \ || empty(bufname)
        \ || match(bufname, '^[a-z]\+://*') isnot# -1
        \ || isdirectory(bufname)
    " Non file buffer
    return
  elseif filereadable(bufname) && filewritable(bufname)
    " File is readable and writeable
    return
  elseif empty(getftype(bufname))
    " if file doesn't exist, we search for a all directories up it's path to
    " see if each one of them is writable, if not, we `return`
    let parent = fnamemodify(bufname, ':p')
    while parent !=# fnamemodify(parent, ':h')
      let parent = fnamemodify(parent, ':h')
      if filewritable(parent) is# 2
        return
      elseif !filereadable(parent) && isdirectory(parent)
        break
      endif
    endwhile
  endif
  let bufnr = str2nr(expand('<abuf>'))
  execute printf(
        \ 'keepalt keepjumps edit %s%s',
        \ g:suda#prefix,
        \ fnamemodify(bufname, ':p'),
        \)
  execute printf('silent! %dbwipeout', bufnr)
endfunction
```

Above fixed the followings

- Use `b:` instead of `s:` so that the variable is removed when the buffer has removed
  - Using `exists()` is enough because users won't touch `b:suda_smart_edit_checked`
- Call `isdirectory` at last because it would invoke OS's file system access
- Move `bufname` and `bufnr` initialization to the line just before requirements while `BufEnter` is invoked often and early return in Vim script is quite important
- Use `expand('<abuf>')` instead of `bufnr('%')` because we use `<afile>` instead of `%`
- Remove comments in a loop block because Vim parse these line repeatedly

Reply:
I agree with every word you've said there. According to my initial testings, everything is fine.